subprojects/core/src/main/java/org/gradle/process/internal/JavaExecHandleBuilder.java

```diff
@@ -311,28 +319,47 @@ protected ArgsEnvironment getEffectiveArgsEnvironment() {
         Map<String, String> environment = getActualEnvironment();
 
         if (hasCommandLineExceedMaxLength(getExecutable(), arguments)) {
-            if (shouldUseEnvironmentVariable(arguments, environment)) {
-                int classPathFlagIndex = arguments.indexOf("-cp");
-                environment.put("CLASSPATH", arguments.get(classPathFlagIndex + 1));
-                arguments.remove(classPathFlagIndex);
-                arguments.remove(classPathFlagIndex);
+            int classPathFlagIndex = arguments.indexOf("-cp");
+            if (shouldTryShorteningCommandLine(arguments)) {
+                if (shouldUseEnvironmentVariable(arguments, environment)) {
+                    environment.put("CLASSPATH", arguments.get(classPathFlagIndex + 1));
+                    arguments.remove(classPathFlagIndex);
+                    arguments.remove(classPathFlagIndex);
+                    LOGGER.info("Gradle is shortening the command line by moving the classpath to the CLASSPATH environment variable.");
+                } else {
+                    try {
+                        File pathingJarFile = File.createTempFile("gradle-javaexec-classpath", ".jar");
+                        List<String> jvmArgs = writePathingJarFile(classpath, pathingJarFile);
+                        arguments.remove(classPathFlagIndex);
+                        arguments.remove(classPathFlagIndex);
+                        arguments.addAll(classPathFlagIndex, jvmArgs);
+                        LOGGER.info("Gradle is shortening the command line by moving the classpath to a pathing JAR.");
```

Comment:
I think this should include the original classpath.  e.g., 

```suggestion
                        LOGGER.info("Shortening Java classpath {} with {}", this.classpath, pathingJarFile);
```

Reply:
Sure, I will make the change.